### ISSUE TITLE
request for a .clone() method

### ISSUE THREAD

**Issue by mxriverlynn** [alessioalex/generic-middleware#14]
I've run into a problem that I should have seen previously, with my Rabbus library.

Whenever you call the `.publish` method on one of my objects, it injects a `.runAfter` into the middleware, to run a specific function after everything is done.

You can probably guess what happens when `.publish` is called more than one... multiple copies of the runAfter are setup, and things break. 

I tried to find a way to add the last function when i initialize the middleware, but I can't do that. if i call `.useAfter` right away, it just adds it as the first middleware function and then subsequent `.use` will end up running after the one that should go last.

Also, due to the closure around a callback function that I am setting up with `.publish`, I have to add the last function with `useAfter`, when `.publish` is called.

So I seem to be in a bit of a bind at the moment... and I think a `.clone` method would solve this for me.

I tried to hack my own clone method into the source in my project, but it's not working right. I added the clone method to the proxy, and modified the Middleware constructor to accept a stack and errorHandlers array as optional arguments. It doesn't seem to work, and I'm not sure what i did wrong.

I'll probably try again if you don't see this before I work in my library again... but I thought I would open this ticket to get the discussion going, first.

**Comment by jokeyrhyme** [comment]
Just to clarify, you want a `.clone()` method that returns a new Middleware object, and this object should have the same middleware functions registered in its stack?

**Comment by mxriverlynn** [comment]
correct - the new instance would have new array instances for the `_stack` and `_errorHandlers`, so that I can independently manipulate the array (add new middleware, etc) without affecting the original.

``` js
var m = new Middleware();
m.use(...);
// m stack count === 1

var m2 = m.clone();
m2.use(...);

// m stack count === 1
// m2 stack count === 2
```

**Comment by alessioalex** [comment]
Hey @derickbailey I'll be looking into this. I'll try to see if we can get around this without needing to clone initially (digging into the use case here -> https://github.com/derickbailey/rabbus/issues/15). If there is a legitimate use case for this, we can add that clone method as well.

**Comment by mxriverlynn** [comment]
@alessioalex thanks - i've been thinking about a way around this, as well and I have an idea of storing the middleware stack / handlers in my own arrays, initially. then when the publish method is called, i create a new instance of the middleware each time, and populate it on demand. 

i'm going to try that out and see how it works. i should be able to do it without any mods to this library... though it might be nice to have a constructor that takes in the 2 arrays. i'll see if the idea works out well enough, first.

**Comment by alessioalex** [comment]
@derickbailey np, I've actually simplified the problem and currently have something in mind that should work quite nicely for this situation. Will update you soon.

**Comment by alessioalex** [comment]
Hey @derickbailey so I've looked at what you are doing in  `rabbus` and simplified the problem down to the following code:

`example.js`: 

``` js
var Something = require('./Something');
var md = new Something();

md.use(function foo(param1, param2, next) {
  console.log('foo called');
  next();
});

md.use(function bar(param1, param2, next) {
  console.log('bar called');
  next();
});

md.request('something1', function handleResult(err, result) {
  console.log('request result 1: ', result);
});

md.request('something2', function handleResult(err, result) {
  console.log('request result 2: ', result);
});

// output show look like this:
/*
foo called
bar called
Something internal logger: requesting the data for something1
Something internal logger: requesting the data for something2
..
(after 100ms) request result 2: bla bla
(after 200ms) request result 1: bla blu
*/
```

The solution consists of 2 tricks: 1) use an `internal` parameter to pass along the arguments of `md.request` as the 1st arg of the middleware 2) hide that parameter and expose a `.use()` function that you control.

The solution is:

`Something.js`:

``` js
var Middleware = require('generic-middleware');

var getSomethingAsync = function(what, cb) {
  var min = 100;
  var max = 500;
  var timeout = Math.random() * (max - min) + min;

  setTimeout(function() {
    cb(null, what.toUpperCase());
  }, timeout);
};

function Something() {
  this.middleware = new Middleware();
  this.middleware.setParams('internal', 'param1', 'param2');

  var that = this;

  this.use = function use(fn) {
    that.middleware.use(function(internal, param1, param2, next) {
      fn(param1, param2, next);
    });
  };

  this._hasFinalFn = false;
};

Something.prototype.request = function(what, cb) {
  if (!this._hasFinalFn) {
    this.middleware.use(function(internal, param1, param2, next) {
      console.log('requesting %s', internal.what);
      getSomethingAsync(internal.what, internal.cb);
    });

    this._hasFinalFn = true;
  }

  this.middleware({ what: what, cb: cb }, 'param1', 'param2');
};

module.exports = Something;
```

So in the `.request()` method we are doing the following things:
- making sure the 'final middleware' fn is added only once (using a boolean)
- using the `internal` object to remember the params passed initially to `.request()` so that those can be used in the 'final middleware'

The `internal` object is being hidden by exposing a `.use()` function that only has the 'public API params' which in turn proxies to the actual middleware stack used.

The output:

``` bash
→ node example.js 
foo called
bar called
requesting something1
foo called
bar called
requesting something2
request result 2:  SOMETHING2
request result 1:  SOMETHING1
```

This approach has the advantage of not having to create a complete new stack with each `request()` call, so imho it's optimal. Let me know what you think and if you think this solves your problem. Also do correct me if I'm wrong in my initial assumptions of the simplified problem.

**Comment by mxriverlynn** [comment]
hmmm... that's an interesting idea. i'm not quite sure if it will work for me, but i'll try it and see. if not, the previous idea i had will most likely work. thanks for digging into the scenario, and for the suggestion, either way! it's always great to get that kind of feedback / help. :)

**Comment by alessioalex** [comment]
Hopefully it will work, as the example was inspired by your original problem in `producer.js`. You are [exporting `.use()` yourself](https://github.com/derickbailey/rabbus/blob/master/rabbus/lib/producer/index.js#L28-L30), [`._request()`](https://github.com/derickbailey/rabbus/blob/master/rabbus/lib/producer/index.js#L78-L95) is the async operation you are doing when calling [`publishMethod()`](https://github.com/derickbailey/rabbus/blob/master/rabbus/lib/producer/index.js#L138) and you're also calling [`.middleware()`](https://github.com/derickbailey/rabbus/blob/master/rabbus/lib/producer/index.js#L141) in your code and not leaving that up to the consumer.

In case it doesn't work though, your idea can be a backup plan as well. I've been initially considering that as well, but ideally you would like to avoid creating a `new Middleware()` with each request, especially considering they are following the same pattern.

Let me know if I can help further, I have setup the project locally and been digging through your `rabbus` producer code so I could hack on stuff as well.

**Comment by mxriverlynn** [comment]
would your suggested solution prevent me from using error handler middleware? 

i'm trying to work on this idea at the moment, and i can't see how this would allow middleware error handlers:

``` js
this.use = function use(fn) {
    that.middleware.use(function(internal, param1, param2, next) {
      fn(param1, param2, next);
    });
  };
```

if i call `foo.use(function(err, param1, param2, next){ ... });`, the `use` function implementation won't pass along the extra parameter for errors. 

i would have to put an if statement, checking against a magic number for the right parameters...

``` js
this.use = function use(fn) {
  if (fn.length === 5) {
    that.middlware.use(function(err, internal, param1, param2, next){
      fn(err, param1, param2, next);
    });
   } else {
    that.middleware.use(function(internal, param1, param2, next) {
      fn(param1, param2, next);
    });
  }
};
```

this is getting a bit ugly, and i don't like having to hard code the parameter length count + 1 like this. 

am i missing something?

**Comment by jokeyrhyme** [comment]
An alternative to using "real" error handlers, is making all of your middleware have an identical function signature, all of them taking an error argument as `arguments[0]`. Just make sure all of your middleware check it before doing anything.

``` js
function myMiddleware (err, blah, foo, next) {
  if (err) { next(err, blah, foo); return; }
  // TODO: normal stuff here
}

function myErrorMiddleware (err, blah, foo, next) {
  if (!err) { next(err, blah, foo); return; }
  // TODO: handle error here
}
```

**Comment by mxriverlynn** [comment]
@jokeyrhyme i'm not fond of that particular pattern for middleware. it muddies the middleware handling... I like the approach of this library following in the footsteps of expressjs, where i work a lot. keeping the error handling and normal middleware functions separate is important to me

**Comment by alessioalex** [comment]
@derickbailey indeed I hadn't taken into consideration the usage of the error handler. I cannot think of a better way of handling that besides what you've shown, maybe just styling it up somewhat:

``` js
this._args = ['param1', 'param2'];
this.middleware.setParams(['internal'].concat(this._args));

this.use = function use(fn) {
  // param1, param2, next
  if (fn.length === this._args.length + 1) {
    that.middleware.use(function(internal, param1, param2, next) {
      fn(param1, param2, next);
    });
  } else {
    that.middlware.use(function(err, internal, param1, param2, next){
      fn(err, param1, param2, next);
    });
  }
};
```

While this may look a bit off, the alternatives are the following:
- have a separate `.addErrorHandler()` method (deviates from the Express API, people are used to `.use()` (pun intended)
- re-create the middleware stack each time as per your idea (less performant)
- the approach suggested by @jokeyrhyme with an error param each time

The code shown above makes the 'magic number' look less magic, and it's only part of the internal implementation of your project, not the public facing API. For the time being I would stay stick with that, since I cannot suggest a better alternative for this specific use case.

Edit: both `Express` and [`generic-middleware`](https://github.com/alessioalex/generic-middleware/blob/master/lib/middleware.js#L25-L31) are doing this when checking for regular middleware vs error handlers.

**Comment by mxriverlynn** [comment]
for the time being, i've run with my idea of rebuilding the middleware each time. it's not optimal, but i got it working quickly. i've also encapsulated it fairly well, so i can change the implementation however i need - hopefully adding your better suggestion as the implementation.

you can see the usage at:
- https://github.com/derickbailey/rabbus/blob/middlewareBuilder/rabbus/lib/consumer/index.js#L18
- https://github.com/derickbailey/rabbus/blob/middlewareBuilder/rabbus/lib/consumer/index.js#L27
- https://github.com/derickbailey/rabbus/blob/middlewareBuilder/rabbus/lib/consumer/index.js#L51-L57

and the MiddlewareBuilder implementation: https://github.com/derickbailey/rabbus/blob/middlewareBuilder/rabbus/lib/middlewareBuilder/index.js

hiding everything behind this MiddlewareBuilder API kept things clean in my use of it... but I am still using a magic number check, as you can see in the implementation. i'm going to try and replace this implementation with your suggestion, hopefully later tonight.

and, i'd love your feedback on what's there, knowing it's not optimal.

**Comment by alessioalex** [comment]
Nice touch with extracting everything into a separate file (middlewareBuilder) since you are reusing the same thing in multiple places. I would say try to implement my suggestion as well and do a simple benchmark (calling `.request()` 10000 times for example) comparing the two approaches.

As for the magic number check, it's unavoidable if you want to use the same function (`.use()`) for both regular and error handler middleware.

Will have a look over stuff tomorrow && continue the discussion.

**Comment by mxriverlynn** [comment]
i added a crude speed test to process 10,000 messages, separating producer and consumer speeds. 

my general numbers, with my current implementation, are sitting around:

```
Totals:
Send Time: 3.001
Send/sec: 3332
Receive Time: 1.542
Receive/sec: 6485
```

this is a pretty typical output set on my laptop. 

... 

i tried to implement your suggesting in my middlewareBuilder and ran into some hiccups due to the differences in parameter counts between objects that use it. will try again, soon.. but i may end up releasing what i have and improving it later.

i'm going to go ahead and close this ticket down as i think it's taken up enough of your time and attention for now. 

thanks again for everything you've done to help guide me down the path with this! i really appreciate it when someone digs in and uncovers the "why" of a question or suggestion, and helps people a better solution than what they were thinking.

**Comment by alessioalex** [comment]
Always happy to help out! Good luck with your project and be sure to post again if you have any issues or suggestions about the module!
